external-crates/move/crates/move-analyzer/editors/code/src/main.ts

```diff
@@ -48,12 +48,56 @@ async function serverVersion(context: Readonly<Context>): Promise<void> {
  * so that you can wait for the activation to complete by await
  */
 export async function activate(extensionContext: Readonly<vscode.ExtensionContext>): Promise<void> {
+    const globalMoveVersionKey = 'move-version';
     const extension = new Extension();
     log.info(`${extension.identifier} version ${extension.version}`);
 
     const configuration = new Configuration();
     log.info(`configuration: ${configuration.toString()}`);
 
+    // VSCode does not provide a hook for install/update extension, and we don't want to attempt
+    // installation of move-analyzer binaries every time an extension is activated (e.g. after
+    // VSCode restart).
+    //
+    // On a happy path (when user does not mock with user settings), we install move-analyzer
+    // whenever the extension itself is installed or upgraded, and skip installation when
+    // the globally stored extension version number does not change. However, even in this
+    // case we want to run the move-analyzer installation procedure if the move-analyzer
+    // is for some reason unavailable (e.g., because the user messed up user settings between
+    // VSCode restarts).
+    //
+    // We also don't want to update the extension version in the global state until we know
+    // that move-analyzer installation succeeded as the global state change is permanent.
+
+    const lastMoveVersion = extensionContext.globalState.get(globalMoveVersionKey);
+    let doInstallBinary: boolean;
+    let updateGlobalExtVersion: boolean;
+    if (lastMoveVersion === null) {
+        // Installation (no global variable set).
+        doInstallBinary = true;
+        updateGlobalExtVersion = true;
+    } else if (lastMoveVersion === extension.version) {
+        // Not an installation or an update (same version as seen before).
+        const serverPathExists = await vscode.workspace.fs.stat(vscode.Uri.file(configuration.serverPath)).then(
+            () => true,
+            () => false,
+        );
+        doInstallBinary = !serverPathExists;
+        updateGlobalExtVersion = false;
+    } else {
+        // Update (different versions).
+        doInstallBinary = true;
+        updateGlobalExtVersion = true;
+    }
+
+    if (doInstallBinary) {
+        const success = await configuration.installServerBinary(extensionContext);
+        if (!success) {
+            return;
+        }
+    }
```

Comment:
I might unnest this, but don't feel super strongly

```suggestion
    if (!doInstallBinary) {
        return;
    }
    
    const success = await configuration.installServerBinary(extensionContext);
    if (!success) {
        return;
    }
```

Reply:
This wouldn't quite work here, as we want to execute code following this conditional even if we skip installation (return only if the attempted installation fails).